evalml/pipelines/components/transformers/preprocessing/time_series_regularizer.py

```diff
@@ -0,0 +1,336 @@
+        if (
+            self.debug_payload["estimated_freq"] is None
+        ):  # If even WW can't infer the frequency
+            raise ValueError(
+                f"The column {self.time_index} does not have a frequency that can be inferred."
+            )
+
+        estimated_freq = self.debug_payload["estimated_freq"]
+        duplicates = self.debug_payload["duplicate_values"]
+        missing = self.debug_payload["missing_values"]
+        extra = self.debug_payload["extra_values"]
+        nan = self.debug_payload["nan_values"]
+
+        self.error_dict = self._identify_indices(
+            self.time_index, X, estimated_freq, duplicates, missing, extra, nan
+        )
+
+        return self
+
+    @staticmethod
+    def _identify_indices(
+        time_index, X, estimated_freq, duplicates, missing, extra, nan
+    ):
+        """Identifies which of the problematic indices is actually misaligned.
+
+        Args:
+            time_index (str): The column name of the datetime values to consider.
+            X (pd.DataFrame): The input training data of shape [n_samples, n_features].
+            estimated_freq (str): The estimated frequency of the `time_index` column.
+            duplicates (list): Payload information regarding the duplicate values.
+            missing (list): Payload information regarding the missing values.
+            extra (list): Payload information regarding the extra values.
+            nan (list): Payload information regarding the nan values.
+
+        Returns:
+            (dict): A dictionary of the duplicate, missing, extra, and misaligned indices and their datetime values.
+        """
+        error_dict = {
+            "duplicate": {},
+            "missing": {},
+            "extra": {},
+            "nan": {},
+            "misaligned": {},
+        }
+
+        # Adds the indices for the consecutive range of missing, duplicate, and extra values
+        for each_missing in missing:
+            # Needed to recreate what the missing datetime values would have been
+            temp_dates = pd.date_range(
+                pd.to_datetime(each_missing["dt"]),
+                freq=estimated_freq,
+                periods=each_missing["range"],
+            )
+            for each_range in range(each_missing["range"]):
+                error_dict["missing"][each_missing["idx"] + each_range] = temp_dates[
+                    each_range
+                ]
+
+        for each_duplicate in duplicates:
+            for each_range in range(each_duplicate["range"]):
+                error_dict["duplicate"][
+                    each_duplicate["idx"] + each_range
+                ] = pd.to_datetime(each_duplicate["dt"])
+
+        for each_extra in extra:
+            for each_range in range(each_extra["range"]):
+                error_dict["extra"][each_extra["idx"] + each_range] = X.iloc[
+                    each_extra["idx"] + each_range
+                ][time_index]
+
+        for each_nan in nan:
```

Comment:
Why do we keep track of the NaNs? We don't really use them else in the implementation?

Reply:
They're mainly needed in testing to ensure that the output X and y lengths are appropriately modified. A side benefit is that it also provides an OS user with a complete view of the information that comes out of the WW payload.